[SPARK-4751] Dynamic allocation in standalone mode

andrewor14 opened the pull request:
Dynamic allocation is a feature that allows a Spark application to scale the number of executors up and down dynamically based on the workload. Support was first introduced in YARN since 1.2, and then extended to Mesos coarse-grained mode recently. Today, it is finally supported in standalone mode as well!

I tested this locally and it works as expected. This is WIP because unit tests are coming.

Commit:
First working implementation

This adds two messages between the AppClient on the driver and
the standalone Master: request and kill. The scheduling on the
Master side handles both applications that explicitly set
`spark.executor.cores` and those that did not.

TODO: clean up shuffle files on application exit and unit tests.

Commit:
Clean up shuffle files after application exits

AmplabJenkins commented:
Merged build triggered.

CodingCat commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -762,6 +789,86 @@ private[master] class Master(
+        if (appInfo.desc.coresPerExecutor.isEmpty) {
+          val numMissingExecutors = appInfo.executorLimit - appInfo.executors.size
+          if (numMissingExecutors > 0) {
+            appInfo.blacklistedWorkers.take(numMissingExecutors).foreach { workerId =>
```
Hi, @andrewor14 , maybe I'm a bit confused here, we add the workers to blacklist to avoid launching the executors immediately after killing them (https://github.com/apache/spark/pull/7532/files#diff-29dffdccd5a7f4c8b496c293e87c8668R836)....but it seems that, if we run one executor per worker, this rule does apply?

CodingCat replied:
I mean when `blacklistedWorkers.length <= numMissingExecutors` ?

andrewor14 replied:
yes, it's OK for `numMissingExecutors > blacklistedWorkers`. In that case we just remove everything from the blacklist. Imagine the case where we have 5 workers (each has 1 executor), then we kill two executors and add their workers to the blacklist. Now if I request 10 extra executors, then 10 > 2 and everything from my blacklist is removed such that we launch executors again, which is what I want.

Does that answer your question?

CodingCat replied:
thanks for answering, but if so, why we have to `blacklist` the workers? Upon I request more resources, the `just-released` workers will be taken again...

my second thought is that even `numMissingExecutors > blacklistedWorkers`, can we ensure that the `numMissingExecutors` executor IDs we take from blacklist are the earliest ones we push into the HashSet? (I think no, since it's a HashSet)

andrewor14 replied:
> thanks for answering, but if so, why we have to blacklist the workers? Upon I request more resources, the just-released workers will be taken again...

Maybe it's not clear in the comments. There are two kinds of applications, ones that set `spark.executor.cores` and ones that didn't. For the applications that didn't set it, each executor just grabs all the available cores on the worker, so each worker can run at most 1 executor.

In this mode, whenever we kill an executor we blacklist its corresponding worker to avoid immediately launching a new executor on that worker. However, there are cases where we *do* want a new executor to be launched after we kill an executor. For example, if we had previously requested 5 total executors, but there is only space for 3, then 2 executors will be waiting. In this case, if we kill 2 existing executors, then 2 new ones should come back up immediately.

> my second thought is that even numMissingExecutors > blacklistedWorkers, can we ensure that the numMissingExecutors executor IDs we take from blacklist are the earliest ones we push into the HashSet? (I think no, since it's a HashSet)

I don't think there are benefits to imposing an ordering on the workers we remove.

CodingCat replied:
I see, so blacklist is just for filtering the workers in the next scheduling rounds until we do request more resources. I thought its semantics is related to "unlike this worker"

Everything is clear now! Thanks!

andrewor14 replied:
Yes, exactly. I've added many more paragraphs of comments in the latest commit to clarify this. Thanks for pointing out the potential source of confusion.

dragos commented on core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala:
```diff
@@ -43,6 +42,18 @@ private[spark] class ApplicationInfo(
+  // A cap on the number of executors this application can have at any given time.
+  // By default, this is infinite. Only after the first allocation request is issued
+  // by the application will this be set to a finite value.
+  @transient var executorLimit: Int = _
```
You could initialize it here to `Int.MaxValue`. `_` is 0, and the comment is misleading, since there is no such thing as infinite values for integers.

andrewor14 replied:
same reason, when this class is deserialized on the Master side, we set this variable to max value. It's not used on the driver side (before it's serialized) so there's no point initializing it here and duplicating the max value in `readObject`.

andrewor14 replied:
(just trying to be consistent with the rest of the variables in this class)

andrewor14 replied:
(see my comment below)

dragos replied:
Oh, I didn't see the `readObject` method. In that case, unfortunately, I see no alternative.

dragos commented on core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala:
```diff
@@ -43,6 +42,18 @@ private[spark] class ApplicationInfo(
+  // at most one executor from this application can be run on each worker. When an executor is
+  // killed, its worker is added to the blacklist to avoid having the master immediately schedule
+  // a new executor on the worker.
+  @transient var blacklistedWorkers: mutable.HashSet[String] = _
```
You should be able to do without *both* a mutable collection *and* a `var`. I'd make this a `val` and initialize it to an empty set right away. Nulls are evil, and I don't see any reason to live with a null.

Probably this `var` should not be public either.

andrewor14 replied:
I think it needs to be a var because we mark it transient, but yes I believe this can be `private[master]`

andrewor14 replied:
I agree that in general we shouldn't have vars for mutable data structures, but this is marked as transient so unfortunately it needs to be a var. This just follows the pattern used in other variables in this class.

I believe the reason why these are marked as transient is because we don't want them to be set on the driver's side (e.g. the driver shouldn't try to blacklist executors). This doesn't have to be the case, but I wanted to minimize the changes in this patch.

andrewor14 replied:
Alright, in the latest commit I have privatized it. I'm going to keep this as a var for the reason I mentioned above unless you feel strongly about it.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Clean up comments and rewrite code for readability

The blacklisting logic is fairly intricate, but there was not
really high level documentation describing why the blacklist
is needed. This commit addresses that.

Commit:
Privatize worker blacklist for cleanliness

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37860/consoleFull) for   PR 7532 at commit [`8352801`](https://github.com/apache/spark/commit/8352801778b50817e0e9a47312c95316fcf4d09d).

SparkQA commented:
[Test build #37860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37860/console) for   PR 7532 at commit [`8352801`](https://github.com/apache/spark/commit/8352801778b50817e0e9a47312c95316fcf4d09d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix style

There's a weird scalastyle bug that thinks there is no space
after the comma in the following:
```
someMethod(name, /* clean up */ true)
```

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37862 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37862/consoleFull) for   PR 7532 at commit [`e282874`](https://github.com/apache/spark/commit/e28287459458805e61f7ae87959d31c7c61fa134).

SparkQA commented:
[Test build #37862 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37862/console) for   PR 7532 at commit [`e282874`](https://github.com/apache/spark/commit/e28287459458805e61f7ae87959d31c7c61fa134).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix MiMa

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37863/consoleFull) for   PR 7532 at commit [`1334e9a`](https://github.com/apache/spark/commit/1334e9ad38ca5f66e8c0b739e273219bf561778f).

SparkQA commented:
[Test build #37863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37863/console) for   PR 7532 at commit [`1334e9a`](https://github.com/apache/spark/commit/1334e9ad38ca5f66e8c0b739e273219bf561778f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

dragos commented on core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala:
```diff
@@ -43,6 +42,18 @@ private[spark] class ApplicationInfo(
+  // at most one executor from this application can be run on each worker. When an executor is
+  // killed, its worker is added to the blacklist to avoid having the master immediately schedule
+  // a new executor on the worker.
+  @transient private var blacklistedWorkers: mutable.HashSet[String] = _
```
I still think it doesn't need to be *both* a `var` and a mutable collection. You could (with minimal changes) make this an `immutable.HashSet`.

My 2c.

dragos commented on core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala:
```diff
@@ -96,6 +109,47 @@ private[spark] class ApplicationInfo(
+   * This is used only if cores per executor is not set.
+   */
+  private[master] def removeFromBlacklist(numWorkers: Int): Unit = {
+    blacklistedWorkers.take(numWorkers).foreach { workerId =>
```
Why not simply `blacklistedWorkers.drop(numWorkers)`?

andrewor14 replied:
drop returns a copy

dragos replied:
Still, there must be a more direct way. Doing a `foreach` for this seems really an anti-pattern. This looks clearer:

```
blacklistedWorkers --= blacklistedWorkers.take(numWorkers)
```

andrewor14 replied:
I removed the blacklist altogether. Never liked it anyway

CodingCat commented on core/src/main/scala/org/apache/spark/deploy/client/AppClient.scala:
```diff
@@ -256,4 +272,33 @@ private[spark] class AppClient(
+   *
+   * @return whether the request is acknowledged.
+   */
+  def requestTotalExecutors(requestedTotal: Int): Boolean = {
```
is it necessary to validate the value of `requestedTotal`, like `>= 0`? though negative numbers does not bring any impact on the correctness of the program (if I understand code correctly)

andrewor14 replied:
this is already done in `CoarseGrainedSchedulerBackend`. We don't need to duplicate that check.

CodingCat commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1387,8 +1374,6 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * This is currently only supported in YARN mode. Return whether the request is received.
```
outdated comments?

andrewor14 replied:
yup, thanks

CodingCat commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1405,8 +1390,6 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   @DeveloperApi
   override def requestExecutors(numAdditionalExecutors: Int): Boolean = {
```
outdated comments for this function? https://github.com/apache/spark/pull/7532/files#diff-364713d7776956cb8b0a771e9b62f82dR1389

CodingCat commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1429,8 +1412,6 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   @DeveloperApi
   override def killExecutors(executorIds: Seq[String]): Boolean = {
```
outdated comments...BTW, how to comment line without the changes?

CodingCat commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -762,6 +789,103 @@ private[master] class Master(
+        // If the application raises the executor limit, then we can launch new executors.
+        // If there are previously blacklisted workers, then we can launch these new executors
+        // by unblacklisting a subset of these workers. For more detail, see `handleKillExecutors`.
+        if (appInfo.desc.coresPerExecutor.isEmpty) {
```
I think we may not need to execute removeFromBlacklist if the application requests less resources than what it has been assigned? (it doesn't change the correctness of the program, since we have this line https://github.com/apache/spark/pull/7532/files#diff-29dffdccd5a7f4c8b496c293e87c8668R598

CodingCat commented:
Hey, @andrewor14 , this patch is a very useful, especially for the users like me, who runs Spark in standalone mode

I left more comments here, most for the outdated comments in SparkContext which says the dynamic allocations are only support in YARN mode

the other question is that 

from the code, I think what will happen if `an application requests (call sc.requestTotalExecutors(x)) less resources than what it has been assigned` is that `it will not get more resources but still runs with the resources in hand`, right?

if so...I'm questioning that if it is a good way to manage the resources, the application has to release the resources by explicitly calling sc.killExecutors()? However...killExecutors is pretty hard to use, I need to get the exact IDs of executors....can we bring a new API like `killExecutors(numExectors: Int)`?

Commit:
Merge branch 'master' of github.com:apache/spark into standalone-da

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/master/Master.scala
	project/MimaExcludes.scala

Commit:
Merge branch 'master' of github.com:apache/spark into standalone-da

Commit:
Simplify logic by removing the worker blacklist

This commit significantly simplifies the logic by using the
executor limit as a cap. Instead of a completely separate code
path that added and removed workers from blacklist, reusing the
executor limit allows us to reason about the behavior of dynamic
allocation more straightforwardly.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38499 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38499/consoleFull) for   PR 7532 at commit [`0a8be79`](https://github.com/apache/spark/commit/0a8be7944c38ecf76ffbf2eb2e3303058fa4968a).

Commit:
Fix comments

SparkQA commented:
[Test build #38499 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38499/console) for   PR 7532 at commit [`0a8be79`](https://github.com/apache/spark/commit/0a8be7944c38ecf76ffbf2eb2e3303058fa4968a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Add tests for scheduling with executor limit

This commit also uncovers a bug where the limit is not applied
correctly for cases where cores per executor is not defined.
The symptom is that every executor launched under this limit
has exactly 1 core. There is a regression test for this case now.

Commit:
Merge branch 'master' of github.com:apache/spark into standalone-da

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38521 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38521/consoleFull) for   PR 7532 at commit [`2e762d6`](https://github.com/apache/spark/commit/2e762d63d6f97b5e6030f1cc30584296107c53d0).

andrewor14 commented:
> from the code, I think what will happen if an application requests (call sc.requestTotalExecutors(x)) less resources than what it has been assigned is that it will not get more resources but still runs with the resources in hand, right?

I don't understand your question. Are you asking what happens if I request more than what my cluster can support? Spark expects the cluster manager to eventually fulfill the request, so once space frees up the Master will launch executors that could not be launched before.

> However...killExecutors is pretty hard to use, because I need to get the exact IDs of executors....can we bring a new API like killExecutors(numExectors: Int)?

That is completely orthogonal to this patch. We can add that in a separate issue if there is demand for it.

SparkQA commented:
[Test build #38521 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38521/console) for   PR 7532 at commit [`2e762d6`](https://github.com/apache/spark/commit/2e762d63d6f97b5e6030f1cc30584296107c53d0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

CodingCat commented:
@andrewor14 for the question, I mean if at the beginning, the application requests 5 executors, but later, it simply calls `sc.requestTotalExecutors(3)` to claim its current demand, the master will not automatically kill the 2 extra exectutors, and the application just runs with 5 executors, right?

CodingCat commented:
if yes, my concern is just on this; this approach requires the user to explicitly call killExecutors to `truly` release resources for others to use...that means, the user always needs to call `requestTotalExecutor` (with a smaller value) and `killExecutors` together...this makes it fishy in terms of the API design....

andrewor14 commented:
@CodingCat `requestTotalExecutor` is never meant to be called by the user. All of this happens if you enable dynamic allocation in the background. The only reason we have `requestTotalExecutors` in the first place is for canceling pending requests, not for killing executors.

Also, as I mentioned, this discussion is completely irrelevant to this patch since the API was added a long time ago. If you wish to suggest something different from the existing design, we should do it through a new issue on JIRA, though I personally don't think it is an important one since it's all internal.

CodingCat commented:
ok, thanks for the explanation

Commit:
Merge branch 'master' of github.com:apache/spark into standalone-da

Conflicts:
	core/src/main/scala/org/apache/spark/SparkContext.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38618 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38618/consoleFull) for   PR 7532 at commit [`24149eb`](https://github.com/apache/spark/commit/24149eb45d593f74d82197e06c45008a1dbdbc21).

SparkQA commented:
[Test build #38618 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38618/console) for   PR 7532 at commit [`24149eb`](https://github.com/apache/spark/commit/24149eb45d593f74d82197e06c45008a1dbdbc21).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`
  * `  case class RequestExecutors(`
  * `class RFormula(override val uid: String) extends Estimator[RFormulaModel] with RFormulaBase `

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix build after merge conflict

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38721/consoleFull) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).

SparkQA commented:
[Test build #38721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38721/console) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1222 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1222/consoleFull) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).

SparkQA commented:
[Test build #1224 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1224/consoleFull) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).

SparkQA commented:
[Test build #1223 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1223/consoleFull) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).

SparkQA commented:
[Test build #1224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1224/console) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

SparkQA commented:
[Test build #1222 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1222/console) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

SparkQA commented:
[Test build #1223 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1223/console) for   PR 7532 at commit [`c0a2c02`](https://github.com/apache/spark/commit/c0a2c02b3f9a98fd15efe10787358ff2fcaade1e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

andrewor14 commented:
@vanzin @nishkamravi2 could you take a look?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -563,32 +570,48 @@ private[master] class Master(
+          true
+        } else {
+          // Otherwise, we should launch a new executor only if we do not exceed the limit
+          assignedExecutors.sum + app.executors.size < app.executorLimit
```
It feels a little weird to insert app limit logic here, since the function seems to be about whether the worker can take a new executor.

nishkamravi2 replied:
Can the if-else be collapsed into: underLimit = assignedExecutors.sum + app.executors.size < app.executorLimit ?

andrewor14 replied:
@vanzin ok I'll separate it out.
@nishkamravi2 no they cannot be collapsed like that, because in the mode where we have one executor per worker, we may have already put 1 core on each worker, then 2 cores and so on.

nishkamravi2 replied:
Yep, makes sense. Looks pretty good to me.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -563,32 +570,48 @@ private[master] class Master(
+    // Keep launching executors until no more workers can accommodate any
+    // more executors, or if we have reached this application's limits
+    var freeWorkers = (0 until numUsable).filter(canLaunchExecutor)
+    while (freeWorkers.nonEmpty) {
```
To continue the previous comment, this seems to be a more natural place to insert a check of whether the application has allocated its max number of executors.

vanzin commented:
Looks ok to me, but my knowledge of standalone mode internals has faded a bit. I just think `schedule()` is a little confusing since it's currently mixing worker and app limits - it would be clearer i.m.o. if those checks were kept separate.

nishkamravi2 commented on core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala:
```diff
@@ -96,6 +101,14 @@ private[spark] class ApplicationInfo(
+   * grab all the available cores on the worker instead.
+   */
+  private[master] def oneExecutorPerWorker(): Boolean = desc.coresPerExecutor.isEmpty
+
```
Is it better of being declared as a local variable in scheduleExecutorsOnWorkers? The decision of translating coresPerExecutor.isEmpty to oneExecutorPerWorker seems to belong to the scheduler, in case we decide to change default behavior in the future.

andrewor14 replied:
ok, sg

nishkamravi2 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -563,32 +570,48 @@ private[master] class Master(
-    // 1 core at a time since we expect a single executor to be launched on each worker
-    val coresPerExecutor = app.desc.coresPerExecutor.getOrElse(1)
+    val coresPerExecutor = app.desc.coresPerExecutor
+    val minCoresPerExecutor = coresPerExecutor.getOrElse(1)
```
Is it cleaner to declare one variable (coresPerExecutor = app.desc.coresPerExecutor.getOrElse(1)) instead of two?

andrewor14 replied:
hm, I thought we used `coresPerExecutor` down there, but maybe not in the latest code. If not we can merge these two.

nishkamravi2 replied:
I think you missed this one

andrewor14 replied:
actually in the latest code we do use `coresPerExecutor` in 2 places. This saves us from duplicating `app.desc.coresPerExecutor` twice. (minor point)

nishkamravi2 replied:
Ah, I see

nishkamravi2 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -804,6 +826,85 @@ private[master] class Master(
+   *
+   * This method assumes the executor limit has already been adjusted downwards through
+   * a separate [[RequestExecutors]] message, such that we do not immediately launch new
+   * executors immediately after the old ones are removed.
```
Nit: redundant 'immediately'

nishkamravi2 commented:
Hey @andrewor14, the scheduling logic and unit tests look good. Would it make sense to add a test for the dynamic allocation behavior?

Commit:
Merge branch 'master' of github.com:apache/spark into standalone-da

Commit:
Address comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

andrewor14 commented:
@vanzin I just looked at the code again and there doesn't seem to be a great way to separate it out. This is because, in the "one executor per worker" case, we actually need to keep scheduling cores on the existing executors even when we're already at the limit. E.g. if the limit is 3 and we have already scheduled 1 core on each worker (1, 1, 1), we should keep scheduling even though we already "have" 3 executors.

Instead, I updated the comment and variable placement to make this easier to read.

SparkQA commented:
[Test build #39276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39276/consoleFull) for   PR 7532 at commit [`accc8f6`](https://github.com/apache/spark/commit/accc8f6f2129272626a276139a8452973fddba0f).

SparkQA commented:
[Test build #39276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39276/console) for   PR 7532 at commit [`accc8f6`](https://github.com/apache/spark/commit/accc8f6f2129272626a276139a8452973fddba0f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`
  * `class MultilayerPerceptronClassifier(override val uid: String)`
  * `class KinesisUtils(object):`
  * `class InitialPositionInStream(object):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

vanzin commented:
> in the "one executor per worker" case, we actually need to keep scheduling cores

I'm not sure I completely understand all the code here, but it seems to me like this case would be covered by just ignoring the requested number of cores in the "one executor per worker" case, and requesting as many cores as the worker has available.

vanzin commented:
But, regardless, don't hold off the patch because of that.

Commit:
Add end-to-end tests for standalone dynamic allocation

This commit is surprisingly difficult because of the complex
interactions between various components in the cluster. Some
modifications to the code were necessary to ensure a strict
ordering of events that the test can assume.

This test uncovered one race condition! If we send an add request
immediately after a kill request, we may not get an executor back
because we never released the cores in the ApplicationInfo. Only
until after the executor is actually killed by the worker do we
release these cores. This is now fixed with a regression test.

Commit:
Merge branch 'master' of github.com:apache/spark into standalone-da

AmplabJenkins commented:
Merged build triggered.

andrewor14 commented:
> I'm not sure I completely understand all the code here, but it seems to me like this case would be covered by just ignoring the requested number of cores in the "one executor per worker" case, and requesting as many cores as the worker has available.

We can't ignore the requested number of cores because we need to support `spark.cores.max` in the spread out case. E.g. we have 5 workers with 10 cores each, if `spark.cores.max = 18` then we should end up with `[6, 6, 6, 5, 5]`. We can't just allocate as all the available cores on the worker, so instead we need to schedule the cores 1 by 1.

AmplabJenkins commented:
Merged build started.

andrewor14 commented:
Alright, as of the latest commit I've added end-to-end tests to this feature in addition to the unit tests. Thanks for the suggestion @nishkamravi2. Given the extensive test coverage and my manual testing I'm am fairly confident about the correctness in this patch.

SparkQA commented:
[Test build #39338 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39338/consoleFull) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).

SparkQA commented:
[Test build #1268 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1268/consoleFull) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).

SparkQA commented:
[Test build #1267 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1267/consoleFull) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).

SparkQA commented:
[Test build #39338 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39338/console) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  public static final class BytesToBytesMapIterator implements Iterator<Location> `
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`
  * `public abstract class UnsafeKeyValueSorter `
  * `abstract class InternalRow extends GenericSpecializedGetters with Serializable `
  * `trait GenericSpecializedGetters extends SpecializedGetters `
  * `case class SortArray(base: Expression, ascendingOrder: Expression)`
  * `case class GetArrayItem(child: Expression, ordinal: Expression)`
  * `case class GetMapValue(child: Expression, key: Expression)`
  * `class ArrayBasedMapData(val keyArray: ArrayData, val valueArray: ArrayData) extends MapData `
  * `class GenericArrayData(array: Array[Any]) extends ArrayData with GenericSpecializedGetters `
  * `abstract class MapData extends Serializable `
  * `public abstract class KVIterator<K, V> `

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1267/console) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #1270 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1270/consoleFull) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).

SparkQA commented:
[Test build #1271 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1271/consoleFull) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39349/consoleFull) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).

SparkQA commented:
[Test build #1270 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1270/console) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

SparkQA commented:
[Test build #1268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1268/console) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1271/console) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #39349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39349/console) for   PR 7532 at commit [`b3c1736`](https://github.com/apache/spark/commit/b3c1736d0672321325d83b406e7e6ac2ec182282).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
Alright, merging into master. After 3 versions dynamic allocation is officially supported on all cluster managers. Woohoo!!

tedyu commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -804,6 +827,87 @@ private[master] class Master(
+          killExecutor(desc)
+        }
+        if (unknown.nonEmpty) {
+          logWarning(s"Application $appId attempted to kill non-existent executors: "
```
Should this be logged at DEBUG level ?

tedyu commented on core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala:
```diff
@@ -152,6 +152,34 @@ private[spark] class SparkDeploySchedulerBackend(
+    Option(client) match {
+      case Some(c) => c.killExecutors(executorIds)
+      case None =>
+        logWarning("Attempted to kill executors before driver fully initialized.")
```
Maybe include executorIds in the log.

mkhaitman commented:
Been testing this merged PR specifically and it looks great so far :). One thing I noticed however is that the TaskSchedulerImpl throws and error once an executor has been lost, even if that's the expected behaviour of using Dynamic Allocation. Would it make more sense to produce a warning as opposed to error in the scenario that the executor is dropped as a result of executor idle timeout?

andrewor14 commented:
Hi @mkhaitman definitely. There's an outstanding JIRA for this [SPARK-4134](https://issues.apache.org/jira/browse/SPARK-4134). This will be addressed in the future hopefully by 1.6.0.
